### PR TITLE
Chart JSON Import support

### PR DESCRIPTION
**Pull request by elatoskinas**
Chart JSON importing support via CSVFile (to be renamed to ChartDataFile)

Corresponds to #92. Refactoring done to accommodate for #57.

### CODE REVIEW

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import com.google.appinventor.client.Ode;
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java:4]: Look into fixing uneven CSV & JSON columns (#100)

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java [hunk 2]
```diff
@@ -69,15 +69,20 @@ public void onFailure(Throwable caught) {
       @Override
       public void onSuccess(List<List<String>> result) {
         // Update rows & columnNames properties
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java]: Update comment

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java [hunk 3]
```diff
@@ -69,15 +69,20 @@ public void onFailure(Throwable caught) {
...
+        columns = result;
 
         if (result.isEmpty()) {
           ErrorReporter.reportError(MESSAGES.emptyFileError());
           return;
         }
 
-        // First row should contain the column names
-        columnNames = result.get(0);
+        for (int i = 0; i < columns.size(); ++i) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java:78]: Comment

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java [hunk 4]
```diff
@@ -148,17 +144,7 @@ public void onSuccess(List<List<String>> result) {
       return column;
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java:143]: Remove TODO with regards to optimization

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java [hunk 5]
```diff
@@ -367,23 +374,58 @@ public String load(long projectId, String fileId) {
     // Load the contents of the specified file
     String result = load(projectId, fileId);
 
-    // Construct an InputStream and a CSVParser for the contents of the file
-    ByteArrayInputStream inputStream = new ByteArrayInputStream(result.getBytes());
-    CsvParser csvParser = new CsvParser(inputStream);
+    List<List<String>> columns = new ArrayList<List<String>>();
 
-    List<List<String>> csvRows = new ArrayList<List<String>>();
+    try {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Commenting

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java [hunk 6]
```diff
@@ -367,23 +374,58 @@ public String load(long projectId, String fileId) {
     // Load the contents of the specified file
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java:375]: Update JavaDoc for method

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java [hunk 7]
```diff
@@ -367,23 +374,58 @@ public String load(long projectId, String fileId) {
...
-      if (!csvParser.hasNext()) { // No more rows exist; break
-        break;
+        columns.add(column);
       }
 
-      // Parse next row and add it to the resulting rows
-      List<String> row = csvParser.next();
-      csvRows.add(row);
+      return columns;
+    } catch (Exception e) {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Replace try-catch with symbol detection for efficiency

File: appinventor/appengine/src/com/google/appinventor/server/properties/json/ServerJsonNumber.java [hunk 8]
```diff
@@ -55,4 +55,8 @@ public short getShort() {
   public String toJson() {
     return "" + value;
   }
+
+  public String toString() {
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/server/properties/json/ServerJsonNumber.java:60]: Add override annotation

File: appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java [hunk 9]
```diff
@@ -231,8 +192,25 @@ private void importFromCSVColumns(ArrayList<YailList> columns, int rows) {
 
             // Add entries to the tuple from all i-th values (i-th row)
             // of the data columns.
-            for (YailList column : columns) {
-                tupleElements.add(column.getString(i));
+            for (int j = 0; j < columns.size(); ++j) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java:193]: Commenting & possibly move to it's own method?

File: appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java [hunk 10]
```diff
@@ -475,6 +453,10 @@ public void setMaximumTimeEntries(int entries) {
      */
     protected abstract YailList getDefaultValues(int size);
 
+    protected String getDefaultValue(int index) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java:471]: Add JavaDoc

File: appinventor/components/src/com/google/appinventor/components/runtime/util/ChartDataSourceUtil.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
+package com.google.appinventor.components.runtime.util;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ChartDataSourceUtil {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/ChartDataSourceUtil.java:11]: JavaDoc for Util class

File: appinventor/components/tests/com/google/appinventor/components/runtime/CSVFileTest.java [hunk 12]
```diff
@@ -316,6 +319,25 @@ public void testGetTwoColumns() {
     testGetColumnsHelper(expected, columns);
   }
 
+//  @Test
```
Comment (elatoskinas) [inline appinventor/components/tests/com/google/appinventor/components/runtime/CSVFileTest.java:417]: Unit testing for CSVFile

File: appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java [hunk 13]
```diff
@@ -284,6 +284,72 @@ public static String getJsonRepresentationIfValueFileName(Object value){
...
+  public static YailList getColumnsFromJSON(String json) {
+    // Parse object from JSON object
+    Object jsonObject = getObjectFromJson(json);
+
+    // Generate columns from parsed JSON object.
+    // Only proceed with column generation logic if the parsed
+    // object is a List. Otherwise, return empty columns
+    List<YailList> resultColumns = new ArrayList<YailList>();
+
+    if (jsonObject instanceof List) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java:309]: Replace if checks and use exceptions (?)

File: appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java [hunk 14]
```diff
@@ -284,6 +284,72 @@ public static String getJsonRepresentationIfValueFileName(Object value){
...
+
+        // Expected type of the entry is a List (key, value pair)
+        // If this is not the case, add nothing to the column elements.
+        if (entry instanceof List) {
+          List listEntry = (List) entry;
+
+          // Add first value as a String to the column elements
+          // List. The first entry should always be a String,
+          // since it is the key.
+          columnElements.add(listEntry.get(0).toString());
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java:325]: Make sure that for JSON #100 is also fixed.

File: appinventor/components/src/com/google/appinventor/components/runtime/util/ChartDataSourceUtil.java (hunk 11 above)
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/ChartDataSourceUtil.java:11]: Unit testing for ChartDataSourceUtil

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java [hunk 15]
```diff
@@ -367,23 +374,58 @@ public String load(long projectId, String fileId) {
...
     String result = load(projectId, fileId);
 
-    // Construct an InputStream and a CSVParser for the contents of the file
-    ByteArrayInputStream inputStream = new ByteArrayInputStream(result.getBytes());
-    CsvParser csvParser = new CsvParser(inputStream);
+    List<List<String>> columns = new ArrayList<List<String>>();
 
-    List<List<String>> csvRows = new ArrayList<List<String>>();
+    try {
+      ServerJsonParser jsonParser = new ServerJsonParser();
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Add MAX_ROWS restriction for JSON parsing

File: appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java (hunk 5 above)
Comment (elatoskinas) [reply appinventor/appengine/src/com/google/appinventor/server/ProjectServiceImpl.java]: Also possibly split CSV & Json parsing to separate private  methods?

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java [hunk 16]
```diff
@@ -19,7 +20,7 @@
   private String sourceFile;
 
   private List<String> columnNames; // First row of the CSV contents
-  private List<List<String>> rows; // Parsed rows of the CSV file
+  private List<List<String>> columns; // Parsed columns of the CSV file
   private Set<CSVFileChangeListener> csvFileChangeListeners;
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java]: Refactor CSVFile to ChartDataFile

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java (hunk 1 above)
Comment (lightingft) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockCSVFile.java]: Handle for JSON as well.

File: appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java [hunk 17]
```diff
@@ -284,6 +284,72 @@ public static String getJsonRepresentationIfValueFileName(Object value){
...
+   * A column consists of the key and all the entries
+   * of the value itself. Each column is a YailList.
+   *
+   * E.g. "x: 5" would be a List (x 5), while
+   * "y: [1,2,3]" would be a List (y 1 2 3)
+   *
+   * @param json  JSON string to parse
+   * @return  YailList of columns, where each entry is a YailList
+   */
+  public static YailList getColumnsFromJSON(String json) {
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/JsonUtil.java]: Unit test getColumnsFromJSON method

File: appinventor/components/src/com/google/appinventor/components/runtime/DataFile.java [hunk 18]
```diff
@@ -227,11 +183,24 @@ public void run() {
                     // Parse InputStream to String
                     final String result = readFromInputStream(inputStream);
 
-                    // Parse rows from the result
-                    rows = CsvUtil.fromCsvTable(result);
+                    // First character is a curly bracket; Assume JSON
+                    if (result.charAt(0) == '{') {
+                        // Parse columns from the result
+                        columns = JsonUtil.getColumnsFromJSON(result);
```
Comment (elatoskinas) [inline appinventor/components/src/com/google/appinventor/components/runtime/DataFile.java]: Fill missing entries with blank entries for consistency

Comment (elatoskinas) [reply appinventor/components/src/com/google/appinventor/components/runtime/DataFile.java]: Added TODO.

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java [hunk 19]
```diff
@@ -5078,14 +5078,14 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
   @Description("Text for Chart type choice 'pie'")
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java:5078]: Change uneven column import to skip missing entries

**Comment by elatoskinas** [comment]
Final checklist of manual testing before merging:
- [x] CSV file
- [x] JSON file
- [x] CSV uneven
- [x] JSON uneven

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartData.java [hunk 20]
```diff
@@ -281,7 +281,7 @@ protected void refreshChart() {
     }
 
     /**
-     * Sets the properties for the Chart Data component.
+     * Sets the properties for the Cha[rt Data component.
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartData.java]: Fix typo

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartDataModel.java [hunk 21]
```diff
@@ -165,8 +164,10 @@ public void setElementsFromColumns(List<List<String>> columns) {
 
                 if (column.size() == 0) { // Empty column; Use default value
                     tuple[j] = getDefaultTupleEntry(i);
-                } else { // Use the column's value
+                } else if (column.size() > i) { // Column has value; Use the column's value (if the column is large enough)
                     tuple[j] = column.get(i);
+                } else {
+                    tuple[j] = "";
```
Comment (elatoskinas) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartDataModel.java:172]: Comment

**Pull request merged**